buffer/src/main/java/io/netty/buffer/api/BufferAllocator.java

```diff
@@ -0,0 +1,167 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+import io.netty.buffer.api.pool.PooledBufferAllocator;
+
+import java.nio.ByteOrder;
+import java.util.function.Supplier;
+
+/**
+ * Interface for allocating {@link Buffer}s.
+ */
+public interface BufferAllocator extends AutoCloseable {
+    /**
+     * Produces a {@link BufferAllocator} that allocates unpooled, on-heap buffers.
+     * On-heap buffers have a {@code byte[]} internally, and their {@linkplain Buffer#nativeAddress() native address}
+     * is zero.
+     * <p>
+     * The concrete {@link Buffer} implementation is chosen by {@link MemoryManager#instance()}.
+     *
+     * @return A non-pooling allocator of on-heap buffers
+     */
+    static BufferAllocator heap() {
+        return new ManagedBufferAllocator(MemoryManager.instance(), false);
+    }
+
+    /**
+     * Produces a {@link BufferAllocator} that allocates unpooled, off-heap buffers.
+     * Off-heap buffers a native memory pointer internally, which can be obtained from their
+     * {@linkplain Buffer#nativeAddress() native address method.
+     * <p>
+     * The concrete {@link Buffer} implementation is chosen by {@link MemoryManager#instance()}.
+     *
+     * @return A non-pooling allocator of on-heap buffers
+     */
+    static BufferAllocator direct() {
+        return new ManagedBufferAllocator(MemoryManager.instance(), true);
+    }
+
+    /**
+     * Produces a pooling {@link BufferAllocator} that allocates and recycles on-heap buffers.
+     * On-heap buffers have a {@code byte[]} internally, and their {@linkplain Buffer#nativeAddress() native address}
+     * is zero.
+     * <p>
+     * The concrete {@link Buffer} implementation is chosen by {@link MemoryManager#instance()}.
+     *
+     * @return A pooling allocator of on-heap buffers
+     */
+    static BufferAllocator pooledHeap() {
+        return new PooledBufferAllocator(MemoryManager.instance(), false);
+    }
+
+    /**
+     * Produces a pooling {@link BufferAllocator} that allocates and recycles off-heap buffers.
+     * Off-heap buffers a native memory pointer internally, which can be obtained from their
+     * {@linkplain Buffer#nativeAddress() native address method.
+     * <p>
+     * The concrete {@link Buffer} implementation is chosen by {@link MemoryManager#instance()}.
+     *
+     * @return A pooling allocator of on-heap buffers
+     */
+    static BufferAllocator pooledDirect() {
+        return new PooledBufferAllocator(MemoryManager.instance(), true);
+    }
+
+    /**
+     * Check that the given {@code size} argument is a valid buffer size, or throw an {@link IllegalArgumentException}.
+     *
+     * @param size The size to check.
+     * @throws IllegalArgumentException if the size is not positive, or if the size is too big (over ~2 GB) for a
+     * buffer to accommodate.
+     */
+    static void checkSize(long size) {
```

Comment:
It's unusual to see long here, given that int is used pretty much everywhere else.  Without looking at the method body (just the doc and signature), it would be awkward to use this.  If I start off with a long, I have to decide to down cast it when I end up asking for the buffer later.   It also raises another question: if checkSize returned fine, would it be gauranteed to be ok when downcast to an int later?

Reply:
Also, it would seem that this method should not be static, since different buffer allocators may have different stringencies.

Reply:
Hmm… maybe this one should just not be part of the public API? It's really only relevant to buffer and allocator implementors.

Reply:
Seems reasonable.   As a user I probably wouldn't call it.